stan/math/prim/fun/ordered_constrain.hpp

```diff
@@ -14,26 +14,22 @@ namespace math {
  * free vector.  The returned constrained vector will have the
  * same dimensionality as the specified free vector.
  *
- * @tparam T type of elements in the vector
+ * @tparam Vec type with a `operator[]` defined.
  * @param x Free vector of scalars.
  * @return Positive, increasing ordered vector.
  * @tparam T Type of scalar.
  */
-template <typename T>
-Eigen::Matrix<T, Eigen::Dynamic, 1> ordered_constrain(
-    const Eigen::Matrix<T, Eigen::Dynamic, 1>& x) {
-  using Eigen::Dynamic;
-  using Eigen::Matrix;
+template <typename Vec, require_vector_like_t<Vec>* = nullptr>
+inline auto ordered_constrain(Vec&& x) {
   using std::exp;
-  using size_type = index_type_t<Matrix<T, Dynamic, 1>>;
 
-  size_type k = x.size();
-  Matrix<T, Dynamic, 1> y(k);
+  auto k = x.size();
```

Comment:
No need for auto here.

Reply:
I thought so too but it looks like users can change this and it's default is

> EIGEN_DEFAULT_DENSE_INDEX_TYPE - the type for column and row indices in matrices, vectors and array (DenseBase::Index). Set to std::ptrdiff_t by default.

So I know it sounds weird but I kind of like auto here. Though casting to int is no biggie (idt) so I'm fine with just calling it int

https://eigen.tuxfamily.org/dox/TopicPreprocessorDirectives.html

Reply:
In that case you can use `Eigen::Index` to avoid the cast.